### ISSUE TITLE
Can shaka player adjust playback speed to keep low latency?

### ISSUE THREAD

**Issue by sobbe81** [shaka-project/shaka-player#4502]
**Have you read the [Tutorials](http://shaka-player-demo.appspot.com/docs/api/tutorial-welcome.html)?**
Yes

**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**
Yes

**What version of Shaka Player are you using?**
4.2.1

**Please ask your question**
We have some low latency linear streams that we are evaluating. We are setting the `lowLatencyMode` parameter and we can see that the latency is much smaller. However, when the network is not good enough the player will run out of buffer and need to re-buffer. When this happens get further away from the live edge and if this happens multiple times the latency becomes bigger and bigger. 

Does shaka player supports playback with a higher speed so the playback is done closer to the live edge again? 

Currently, we have only tested this scenario with DASH but I guess the same issue and question applies to HLS as well.

**Comment by joeyparrish** [comment]
Currently, Shaka Player does not do this automatically, but you can do it at the application level.  For example (untested, pseudo-code):

```js
video.addEventListener('timeupdate', () => {
  const seekRange = player.seekRange();
  const distanceToEdge = seekRange.end - video.currentTime;
  if (distanceToEdge > SOME_THRESHOLD) {
    player.trickPlay(1.02);  // Should not be noticeable to the user, but will catch up eventually.
  } else {
    player.trickPlay(1);
  }
});
```

You might want multiple thresholds to catch up more or less depending on distance, or some other checks to avoid unnecessary calls to trickMode(), but this is the rough idea of how you would select a "catch up" rate.

Does this help?

**Comment by joeyparrish** [comment]
FWIW, I'm in favor of adding it to Shaka, but I wanted to give a quick app-level snippet that could help in the meantime.

I haven't seen dash.js's version of this, but I recently published a personal project that involves (among other things) syncing two or more media elements.  It may be a bit more complex than we need in Shaka, but it could be adapted easily enough to sync one media element against a target offset from live.

https://github.com/joeyparrish/upfish/blob/bc6a45409ebc43e8f766c21a000ae0fc9a989921/src/upfish.js#L349-L412

The meat of it isn't much, and isn't much different from the snippet I gave above.  But just to be clear on rights, if anybody decides to use any part of that logic for this, you have my express permission as the author of that GPLv3 code to relicense it to this project under Apache 2.
